core/src/main/java/org/elasticsearch/cluster/routing/allocation/AllocateUnassignedDecision.java

```diff
@@ -109,13 +111,7 @@ public AllocateUnassignedDecision(StreamInput in) throws IOException {
         allocationStatus = in.readOptionalWriteable(AllocationStatus::readFrom);
         assignedNode = in.readOptionalWriteable(DiscoveryNode::new);
         allocationId = in.readOptionalString();
-
-        Map<String, NodeAllocationResult> nodeDecisions = null;
-        if (in.readBoolean()) {
-            nodeDecisions = in.readMap(StreamInput::readString, NodeAllocationResult::new);
-        }
-        this.nodeDecisions = (nodeDecisions != null) ? Collections.unmodifiableMap(nodeDecisions) : null;
-
+        nodeDecisions = in.readBoolean() ? sortNodeDecisions(in.readMap(StreamInput::readString, NodeAllocationResult::new)) : null;
```

Comment:
is sorting needed when we read the map? Can we assume it's sorted already and use a LinkedHashMap? This could be done by generalizing readMap so that it also takes a Map supplier.